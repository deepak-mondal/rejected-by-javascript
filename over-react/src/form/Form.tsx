import { useState } from "react";

const Form = () => {
	const [inputString, setInputString] = useState<string>("");
	const [list, setList] = useState<string[]>([]);

	const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault();
		if (inputString.trim().length < 1) {
			setInputString("");
			return;
		}
		setList([...list, inputString]);
		setInputString("");
	};

	return (
		<form onSubmit={handleSubmit} data-testid="form">
			<input aria-label="enter-todos" onChange={(e) => setInputString(e.target.value)} value={inputString} data-testid="input-box"/>
			<button type="submit" data-testid="button">Add</button>
			{list.map((item, index) => {
				return <div key={"todo" + index} data-testid="list-items">{item}</div>;
			})}
			<select data-testid="select-box" aria-label="select-todos">
				<option data-testid="option">text 1</option>
				<option data-testid="option">text 2</option>
				<option data-testid="option">text 3</option>
				<option data-testid="option">text 4</option>
			</select>
		</form>
	);
};

export default Form;
