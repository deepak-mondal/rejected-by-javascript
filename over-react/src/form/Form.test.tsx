import Form from "./Form";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

describe("Form tests", () => {
	afterEach(() => {
		cleanup();
	});

	it("should render all components", async () => {
		render(<Form />);

		const formWrapper = screen.getByTestId("form");
		const inputBox = screen.getByTestId("input-box");
		const button = screen.getByTestId("button");
		const list = screen.queryAllByTestId("list-items");

		expect(formWrapper).toBeInTheDocument();
		expect(inputBox).toBeInTheDocument();
		expect(button).toBeInTheDocument();
		expect(list.length).toBe(0);
	});

	it("should be able to type in the input box", () => {
		const { container } = render(<Form />);

		const inputBox = screen.getByTestId("input-box");
		const button = screen.getByTestId("button");

		expect(inputBox).toBeInTheDocument();
		expect(button).toBeInTheDocument();

		fireEvent.change(inputBox, { target: { value: "test" } });
		expect(inputBox).toHaveValue("test");

		expect(container).toMatchSnapshot();
	});

	it("select test", () => {
		render(<Form />);
		const select = screen.getByTestId("select-box");

		expect(select).toBeInTheDocument();
		fireEvent.click(select);
		const options = screen.getAllByTestId("option");
		const [selectedOption] = options;
		fireEvent.change(select, { target: { value: selectedOption.innerHTML } });
		expect(select).toHaveValue(selectedOption.innerHTML);
	});
});
