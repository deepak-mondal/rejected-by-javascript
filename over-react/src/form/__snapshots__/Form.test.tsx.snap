// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form tests > should be able to type in the input box 1`] = `
<div>
  <form
    data-testid="form"
  >
    <input
      aria-label="enter-todos"
      data-testid="input-box"
      value="test"
    />
    <button
      data-testid="button"
      type="submit"
    >
      Add
    </button>
    <select
      aria-label="select-todos"
      data-testid="select-box"
    >
      <option
        data-testid="option"
      >
        text 1
      </option>
      <option
        data-testid="option"
      >
        text 2
      </option>
      <option
        data-testid="option"
      >
        text 3
      </option>
      <option
        data-testid="option"
      >
        text 4
      </option>
    </select>
  </form>
</div>
`;
